### PR TITLE
Openidc

### PR DESCRIPTION
**Pull request by puiterwijk**

### CODE REVIEW

File: fedora/client/openidcclient.py [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+    def _idp_url(self, endpoint):
+        return self.idp + endpoint
+
+    def _get_token_info(self, token):
+        self.debug('Retrieving token info')
+        resp = requests.post(self._idp_url('TokenInfo'),
+                             data={'client_id': self.client_id,
+                                   'client_secret': self.client_secret,
+                                   'token': token['access_token']})
+        # TODO: Be a bit more stable in handling if we get a non-JSON object
```
Comment (jeremycline) [inline fedora/client/openidcclient.py]: I know this is a work-in-progress that you haven't tested and kindly opened for me, but here's some drive-by feedback ✈️ 

I'd recommend doing something like
```
# Throw an HTTPError-based exception for a HTTP 400+ code
resp.raise_for_status()
try:
    return resp.json()
except JSONDecodeError:
   raise SomeExceptionThatTellsUsersTheServerShouldHaveGivenUsJSON
```

**Comment by ralphbean** [comment]
Just FYI, module-build-service is blocking on this PR over here:  https://pagure.io/fm-orchestrator/issue/268

File: fedora/client/openidcclient.py [hunk 2]
```diff
@@ -0,0 +1,385 @@
+#!/usr/bin/env python2 -tt
```
Comment (igor-raits) [inline fedora/client/openidcclient.py]: bad idea to have shebangs (especially python2's ;))

File: testopenidc.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+from fedora.client.openidcclient import OpenIDCBaseClient
+
+import logging
+logging.basicConfig(level=logging.DEBUG)
+
+inst = OpenIDCBaseClient('http://myapp.com', 'myapp', id_provider='http://172.17.0.2:8080/openidc/')
+
+print inst.get_token(['ipsilon'], False)
```
Comment (igor-raits) [inline testopenidc.py]: shouldn't it be print functions?

File: fedora/client/openidcclient.py [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+            if 'error' in kv:
+                self.debug('Error code returned: %s (%s)'
+                           % (kv['error'], kv.get('error_description')))
+                self.__retrieved_code = False
+            else:
+                self.__retrieved_code = kv['code']
+
+            # Just return a message
+            out = StringIO()
+            print >>out, "You can close this window and return to the CLI"
```
Comment (igor-raits) [inline fedora/client/openidcclient.py]: py2.. meh

**Review by igor-raits — COMMENTED** [review]
too much py2-only stuff =(

File: fedora/client/openidcclient.py [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+        if len(scopes) > 0:
+            query['scope'] = 'openid %s' % ' '.join(scopes)
+        else:
+            query['scope'] = 'openid'
+        query['response_type'] = 'code'
+        query['client_id'] = self.client_id
+        query['redirect_uri'] = return_uri
+        query['state'] = 'ignored'
+        query['response_mode'] = 'query'
+        query = urllib.urlencode(query)
```
Comment (igor-raits) [inline fedora/client/openidcclient.py]: not gonna work on py3.. it's in urllib.parse there. you can use six.moves to get urllib_parse

File: fedora/client/openidcclient.py (hunk 1 above)
Comment (maxamillion) [reply fedora/client/openidcclient.py]: I like @jeremycline's suggestion 👍

**Comment by maxamillion** [comment]
LGTM 👍

**Pull request merged**